transport/http_transport.go

```diff
@@ -0,0 +1,467 @@
+
+	var err error
+	t.listener, err = net.Listen("tcp", address)
+	if err != nil {
+		return fmt.Errorf("while starting HTTP listener: %w", err)
+	}
+
+	t.server = &http.Server{
+		Handler: mux,
+	}
+
+	t.wg.Add(1)
+	go func() {
+		t.opts.Logger.Info(fmt.Sprintf("Listening on %s ....", address))
+		if err := t.server.Serve(t.listener); err != nil {
+			if !errors.Is(err, http.ErrServerClosed) {
+				t.opts.Logger.Error("while starting HTTP server", "err", err)
+			}
+		}
+		t.wg.Done()
+	}()
+
+	// Ensure server is accepting connections before returning
+	return waitForConnect(ctx, t.listener.Addr().String(), nil)
+}
+
+// ShutdownServer shuts down the server started when calling SpawnServer()
+func (t *HttpTransport) ShutdownServer(ctx context.Context) error {
+	if err := t.server.Shutdown(ctx); err != nil {
+		return err
+	}
+	t.wg.Wait()
+	return nil
+}
+
+// ListenAddress returns the address the server is listening on after calling SpawnServer().
+func (t *HttpTransport) ListenAddress() string {
+	return t.listener.Addr().String()
+}
+
+// ServeHTTP handles all incoming HTTP requests received by the server spawned by SpawnServer()
+func (t *HttpTransport) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if t.instance == nil {
+		panic("groupcache instance is nil; you must register an instance by calling HttpTransport.Register()")
+	}
+
+	if !strings.HasPrefix(r.URL.Path, t.opts.BasePath) {
+		panic("HTTPPool serving unexpected path: " + r.URL.Path)
+	}
+
+	parts := strings.SplitN(r.URL.Path[len(t.opts.BasePath):], "/", 2)
+	if len(parts) != 2 {
+		http.Error(w, "bad request", http.StatusBadRequest)
+		return
+	}
+	groupName := parts[0]
+	key := parts[1]
+
+	type transportMethods interface {
+		Get(ctx context.Context, key string, dest Sink) error
+		LocalSet(string, []byte, time.Time)
+		LocalRemove(string)
+	}
+
+	// Fetch the value for this group/key.
+	group := t.instance.GetGroup(groupName).(transportMethods)
+	if group == nil {
+		http.Error(w, "no such group: "+groupName, http.StatusNotFound)
+		return
+	}
+	var ctx context.Context
+	if t.opts.Context != nil {
+		ctx = t.opts.Context(r)
+	} else {
+		ctx = r.Context()
+	}
+
+	// Delete the key and return 200
+	if r.Method == http.MethodDelete {
+		group.LocalRemove(key)
+		return
+	}
+
+	// The read the body and set the key value
+	if r.Method == http.MethodPut {
+		defer r.Body.Close()
+		b := bufferPool.Get().(*bytes.Buffer)
+		b.Reset()
+		defer bufferPool.Put(b)
+		_, err := io.Copy(b, r.Body)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		var out pb.SetRequest
+		err = proto.Unmarshal(b.Bytes(), &out)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		var expire time.Time
+		if out.Expire != nil && *out.Expire != 0 {
+			expire = time.Unix(*out.Expire/int64(time.Second), *out.Expire%int64(time.Second))
+		}
+
+		group.LocalSet(*out.Key, out.Value, expire)
+		return
+	}
+
```

Comment:
Do you think it's worth explicitly checking for a GET & then returning a HTTP 405 for anything else? I assume the transport behaves as expected but unintentional calls to the endpoint with all the other verbs seems like they would prompt undesired Get calls?